src/core/filesystem/mod.rs

```diff
@@ -0,0 +1,471 @@
+#[cfg(not(target_os = "android"))]
+pub fn get_engine_base() -> EngineBase { () }
+
+// Filesystem assumptions:
+// 1. Url can be arbitrary and doesn't have to contain any names
+// 2. We can have access to a folder via url, but we can't assume it's a path
+// 3. We can create new files in a folder and get an url to that file, and that url can be arbitrary
+// 4. We can have access to file data only, without any info about it's location, folder, parent or other files in the same folder
+
+// Testing:
+// - Dragging video file
+// - Dragging .gyroflow file
+// - Dragging .gyroflow preset
+// - Opening video file
+// - Opening .gyroflow file
+// - Opening .gyroflow preset
+// - Dragging lens profile
+// - Opening lens profile
+// - Dragging gyro data
+// - Opening gyro data
+// - Changing output path
+// - Selecting output folder
+// - Rendering to existing file
+// - Adding file to queue
+// - Adding multiple files to queue
+// - Exporting lens profile
+// - Dragging preset to render queue
+// - Rendering with project file
+// - Saving project file
+// - Opening video with project file next to it
+// - Merging mp4
+// - Detecting mp4 sequence
+// - Rendering r3d with conversion
+// - Using CLI
+
+#[derive(thiserror::Error, Debug)]
+pub enum FilesystemError {
+    #[error("Invalid url {0:?}")]
+    InvalidUrl((String, url::ParseError)),
+
+    #[error("Not a file url {0}")]
+    NotAFile(String),
+
+    #[error("Not a folder url {0}")]
+    NotAFolder(String),
+
+    #[error("Path doesn't have a parent {0}")]
+    NoParent(String),
+
+    #[error("Invalid path {0}")]
+    InvalidPath(String),
+
+    #[error("Invalid file descriptor {0}")]
+    InvalidFD(i32),
+
+    #[error("Unknown error")]
+    Unknown,
+
+    #[error("IO error")]
+    IOError(#[from] std::io::Error),
+
+    #[error("String error")]
+    Utf8Error(#[from] std::str::Utf8Error),
+
+    #[cfg(target_os = "android")]
+    #[error("Java exception")]
+    JavaException(#[from] jni::errors::Error)
+}
+#[macro_export]
+macro_rules! function_name {
+    () => {{
+        fn f() {}
+        fn type_name_of<T>(_: T) -> &'static str { std::any::type_name::<T>() }
+        type_name_of(f).rsplit("::").find(|&part| part != "f" && part != "{{closure}}").expect("Short function name")
+    }};
+}
+#[macro_export]
+macro_rules! dbg_call {
+    ($($param:ident )* $(-> $ret:expr)*) => {{
+        let mut dbg_str = format!("{}", function_name!());
+        $(
+            dbg_str.push_str(&format!(" | {}: {}", stringify!($param), $param));
+        )*
+        $(
+            dbg_str.push_str(" -> ");
+            dbg_str.push_str(&format!("{:?}", $ret));
+        )*
+        log::debug!("{}", dbg_str);
+    }};
+}
+macro_rules! result {
+    ($result:expr $(, $param:ident)*) => {{
+        let ret = $result;
+        dbg_call!($($param )* -> ret);
+        match ret {
+            Ok(x) => x,
+            Err(e) => panic!("{e:?}") // TODO: log::error!("{e:?}"); String::new()

```

Comment:
If there's any unexpected filesystem error, the program will panic and crash. This is intentional for now, because we need to know if something goes wrong. It will be changed to log::error! later